### PR TITLE
A few perf. improvements for state handling

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

- Make sure that we don't actually allocate state in tracer mode (e.g., in library mode to detect mid-circuit measurements).

- Skip state vector validation by default.

- Add some more timing traces to custatevec backend for future analysis.

### CODE REVIEW

File: runtime/nvqir/CircuitSimulator.h [hunk 1]
```diff
@@ -807,7 +807,8 @@ class CircuitSimulatorBase : public CircuitSimulator {
   std::size_t allocateQubit() override {
     // Get a new qubit index
     auto newIdx = tracker.getNextIndex();
-
+    const bool inTraceMode =
+        (executionContext && executionContext->name == "tracer");
```
Comment (schweitzpgi) [inline runtime/nvqir/CircuitSimulator.h:811]: Looks like you could make this a small `static` function in this file.

File: runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu [hunk 2]
```diff
@@ -298,9 +298,14 @@ protected:
       // FIXME handle case where pointer is a device pointer
 
       // First allocation, so just set the user provided data here
-      HANDLE_CUDA_ERROR(cudaMemcpy(deviceStateVector, state,
-                                   stateDimension * sizeof(CudaDataType),
-                                   cudaMemcpyHostToDevice));
+      {
```
Comment (schweitzpgi) [inline runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu:301]: Do you really need this block (`{ ... }`)?

File: runtime/nvqir/CircuitSimulator.h (hunk 1 above)
Comment (1tnguyen) [reply runtime/nvqir/CircuitSimulator.h]: Indeed, fixed in https://github.com/NVIDIA/cuda-quantum/pull/1607/commits/96b1112502e95b34cec0c9271b05ffe4035667cf

File: runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu (hunk 2 above)
Comment (1tnguyen) [reply runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu]: No, we don't. Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1607/commits/96b1112502e95b34cec0c9271b05ffe4035667cf

File: runtime/cudaq/qis/qvector.h [hunk 3]
```diff
@@ -60,6 +61,7 @@ class qvector {
                          : simulation_precision::fp64;
     getExecutionManager()->initializeState(targets, vector.data(), precision);
   }
+  qvector(const std::vector<complex> &vector) : qvector(vector, false){};
```
Comment (schweitzpgi) [inline runtime/cudaq/qis/qvector.h:64]:
```suggestion
  qvector(const std::vector<complex> &vector) : qvector(vector, /*validate=*/false){};
```

**Review by schweitzpgi — APPROVED** [review]
LGTM. (Small nit.)

**Pull request merged**